src/appleseed/renderer/kernel/shading/closures.cpp

```diff
@@ -2456,17 +2442,47 @@ Color3f process_background_tree(const OSL::ClosureColor* ci)
     return do_process_closure_id_tree(ci, BackgroundID);
 }
 
-void process_matte_tree(const OSL::ClosureColor* ci, foundation::Color3f& matte_color, Alpha& matte_alpha)
+void process_matte_tree(const OSL::ClosureColor* closure, foundation::Color3f& matte_color, Alpha& matte_alpha)
 {
-    if (ci && ci->id == MatteSurfaceID)
+    if (closure)
     {
-        const OSL::ClosureComponent* c = reinterpret_cast<const OSL::ClosureComponent*>(ci);
-
-        if (c->id == MatteSurfaceID)
+        switch (closure->id)
         {
-            const MatteClosure::Params* p = static_cast<const MatteClosure::Params*>(c->data());
-            matte_alpha.set(p->matte_alpha);
-            matte_color = p->matte_color;
+          case OSL::ClosureColor::MUL:
+            {
+                const OSL::ClosureMul* c = reinterpret_cast<const OSL::ClosureMul*>(closure);
+                foundation::Color3f color_a;
+                Alpha alpha_a;
+                process_matte_tree(c->closure, color_a, alpha_a);
+                matte_color = Color3f(c->weight) * color_a;
+                matte_alpha.set(saturate(luminance(Color3f((c->weight) * alpha_a[0]))));
+            }
+            break;
+
+          case OSL::ClosureColor::ADD:
+            {
+                const OSL::ClosureAdd* c = reinterpret_cast<const OSL::ClosureAdd*>(closure);
+                foundation::Color3f color_a, color_b;
+                Alpha alpha_a, alpha_b;
+                process_matte_tree(c->closureA, color_a, alpha_a);
+                process_matte_tree(c->closureB, color_b, alpha_b);
+                matte_color = color_a + color_b;
+                matte_alpha.set(alpha_a[0] + alpha_b[0]);
+            }
+            break;
+
+          default:
+            {
+                const OSL::ClosureComponent* c = reinterpret_cast<const OSL::ClosureComponent*>(closure);
+
+                if (c->id == MatteSurfaceID)
+                {
+                    const MatteClosure::Params* p = static_cast<const MatteClosure::Params*>(c->data());
+                    matte_alpha.set(p->matte_alpha);
+                    matte_color = p->matte_color;
+                }
+            }
```

Comment:
We need to also check for HoldoutID here and do matte_alpha = 0, matte_color = 0 if found.

Reply:
Hi @est77 , I've made this change and then I changed so any shader except Matte will return zero color and zero alpha. It made blending matte with normal shaders to not fail. Now blending of two mattes works correctly, but blending with normal shader looks as if matte is blended with 0.

Reply:
We don't need the else part. Just ignore and skip anything that is not a MatteID.